### ISSUE TITLE
rocksdb::CompactRangeOptions.change_level is not compatible with level_compaction_dynamic_level_bytes

### ISSUE THREAD

**Issue by hotpxl** [apache/kvrocks#2601]
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/kvrocks/issues) and found no similar issues.

### Version

6b23c5ac2cf17558d167a3305c8f10ecb944637f

### Minimal reproduce step

Default configuration but with `stats_dump_period_sec` set (so you can see compaction stats) , `redis-cli set a b`, `redis-cli compact`.

### What did you expect to see?

Because of how `level_compaction_dynamic_level_bytes` works, there should be only 1 compaction to L6 (the default bottom-most layer)

### What did you see instead?

1 compaction on each layer and 2 compactions to L6

### Anything Else?

I instrumented rocksdb code and this is because `rocksdb::CompactRangeOptions.change_level` is not compatible with `level_compaction_dynamic_level_bytes`.

When `level_compaction_dynamic_level_bytes` is set, `set a b` and `compact` will not compact to L1, it'll jump directly to the last layer (L6). And when `change_level` is set, it then moves this SST file from L6 back to L1.

Now when another periodic compaction comes along, it tries to move this file from L1->L2, L2->L3, ... This is why every layer has 1 compaction and the last layer has 2 compactions.

If you run a manual compaction at this time, since `change_level` is set, it'll move the SST file to L1 again, resulting in the same churn happening.

Obviously this is quite efficiently since these 2 options are trying to undo each other's work..

### Are you willing to submit a PR?

- [x] I'm willing to submit a PR!

**Comment by mapleFU** [comment]
🤔so should we disable `CompactRangeOptions::change_level` or should hear rocksdb's advice on this? I found https://groups.google.com/g/rocksdb/c/wSEk4jMRui0/m/cDwPabfQAQAJ but not has correspond respond

**Comment by hotpxl** [comment]
yeah i think we should just disable `change_level`.. i don't think rocksdb is actively looking at it

**Comment by hotpxl** [comment]
what do you think? i'm happy to make this change

**Comment by git-hulk** [comment]
@hotpxl Thanks for your investigation, and I'm good with disabling the `change_level`. cc @mapleFU

**Comment by mapleFU** [comment]
Sorry for delaying, I'm also good to disable this. I'm not so familiar with how `change_level` works, I'll take a careful round about this part of code this weekend

**Comment by git-hulk** [comment]
> Sorry for delaying, I'm also good to disable this. I'm not so familiar with how `change_level` works, I'll take a careful round about this part of code this weekend

The rocksdb's wiki[1] has a clear explanation about this option. The main goal is to try best to do the compactoin migration to minimum level. And from the implementation, it requires the source level is empty after moving(cannot partially move files in the level).


> CompactRangeOptions::change_level,CompactRangeOptions::target_level Together, these options control the level where the compacted files will be placed. If target_level is -1, the compacted files will be moved to the minimum level whose computed max_bytes is still large enough to hold the files. Intermediate levels must be empty. For example, if the files were initially compacted to L5 and L2 is the minimum level large enough to hold the files, they will be placed in L2 if L3 and L4 are empty or in L4 if L3 is non-empty. If target_level is positive, the compacted files will be placed in that level provided intermediate levels are empty. If any any of the intermediate levels are not empty, the compacted files will be left where they are.


[1] https://github.com/facebook/rocksdb/wiki/Manual-Compaction

**Comment by mapleFU** [comment]
Yeah, but this not explain how it affected by dynamic_level_bytes...

**Comment by hotpxl** [comment]
> Yeah, but this not explain how it affected by dynamic_level_bytes...

yeah in general i feel that rocksdb doesn't do a good job of explaining how multiple things work together. in this case it's a very inefficient interaction between those 2 options, but luckily no correctness issue.

i have found some other instances where rocksdb isn't exactly wrong, but isn't making it clear and unconfusing for the user as well. i'll make some PRs soon

**Comment by hotpxl** [comment]
Yeah I spent some more effort looking into why it didn't work in RocksDB and created an issue there with my findings. Let's see if they respond anytime soon but I think disabling it here in Kvrocks is the best approach forward, which I'll get back to!
